File: src/player/order.rs
//! Queue ordering policy (pure functions).
//!
//! Port of `web/player-core/src/order.ts` (BSD-3-Clause). Every function is
//! deterministic; shuffle takes an injected RNG so hosts/tests seed
//! randomness (purity law: no ambient globals).

use super::types::RepeatMode;

/// Shuffle RNG contract: a value in `[0, 1)`.
///
/// A host RNG that returns exactly `1.0` would select an out-of-range index
/// in the TypeScript implementation; this port clamps defensively so a
/// misbehaving host cannot panic or corrupt the permutation.
pub type Rng = dyn FnMut() -> f64;

/// Fisher–Yates shuffle over `upcoming`. Returns a new vector; the input is
/// untouched.
pub fn shuffle_order<T: Clone>(upcoming: &[T], rng: &mut Rng) -> Vec<T> {
    let mut out = upcoming.to_vec();
    if out.is_empty() {
        return out;
    }
    let mut i = out.len() - 1;
    while i > 0 {
        let raw = rng();
        let scaled = raw * (i as f64 + 1.0);
        let j = if scaled.is_finite() {
            (scaled.floor() as i64).clamp(0, i as i64) as usize
        } else {
            0
        };
        out.swap(i, j);
        i -= 1;
    }
    out
}

/// The index of the next item under the given repeat mode.
///
/// `off`/`one` return `None` at the end (`one` repeats via a player reload,
/// not here); `all` wraps to 0.
pub fn next_index_under_repeat(cursor: i64, length: usize, repeat: RepeatMode) -> Option<usize> {
    if length == 0 {
        return None;
    }
    if cursor + 1 < length as i64 {
        return Some((cursor + 1) as usize);
    }
    if repeat == RepeatMode::All {
        Some(0)
    } else {
        None
    }
}

File: web/player-core/src/order.ts
// Copyright (c) 2026, The MusicPack Development Team
// SPDX-License-Identifier: BSD-3-Clause

// Queue ordering policy (player-core M3 scaffolding; consumed by M5).
//
// Pure functions over the queue's canonical items. Shuffle never destroys
// the canonical order: it produces a presentation ORDER (a permutation) that
// QueueModel applies on top. Determinism: every function takes an injectable
// RNG so tests (and future hosts) can seed randomness — purity law 2.

import type { PlaybackItem } from './types';

export type RepeatMode = 'off' | 'one' | 'all';

export type Rng = () => number;

/** Fisher–Yates over `upcoming` using `rng`. Returns a new array; inputs
 *  are untouched. */
export function shuffleOrder<T>(upcoming: readonly T[], rng: Rng): T[] {
  const out = [...upcoming];
  for (let i = out.length - 1 > 0 ? out.length - 1 : 0; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    const a = out[i] as T;
    const b = out[j] as T;
    out[i] = b;
    out[j] = a;
  }
  return out;
}

/** The index of the next item under the given repeat mode.
 *  - off/one: null at the end ('one' repeats via reload in Player, not here)
 *  - all: wraps to 0 */
export function nextIndexUnderRepeat(
  cursor: number,
  length: number,
  repeat: RepeatMode,
): number | null {
  if (length === 0) return null;
  if (cursor + 1 < length) return cursor + 1;
  return repeat === 'all' ? 0 : null;
}

/** Identity helper: stable per-queue-entry id when a host has none. */
export function itemKey(item: PlaybackItem): string {
  return `${item.id}@${item.trackId}`;
}
